### PR TITLE
Add support for SSL

### PR DESCRIPTION
**Pull request by water-a**
This adds support for SSL connections.

You can provide SSL options in the db config:
Providing certificate to verify SSL/TLS
```
ssl: {
  host: "insert IP of remote database",
  port: 5432,
  ca: ["insert ca cert here"],
}
```

Ignoring SSL/TLS errors; allows for MITM attacks to occur
or 

```
"ssl": {
  "rejectUnauthorized": false
}
```

or 

Checks local certificate
```
"ssl": true
```

### CODE REVIEW

File: packages/wire/src/queue.ts [hunk 1]
```diff
@@ -27,19 +28,83 @@ export class AsyncQueue {
...
+            case 'S':
+              break;
+            case 'N':
+              this.socket.end();
+              throw new Error('The server does not support SSL connections');
+            default:
+              this.socket.end();
+              throw new Error(
+                'There was an error establishing an SSL connection',
+              );
```
Comment (adelsz) [inline packages/wire/src/queue.ts:77]: Can this message be moved here?
https://github.com/adelsz/pgtyped/blob/master/packages/wire/src/messages.ts

File: packages/wire/src/queue.ts [hunk 2]
```diff
@@ -27,19 +28,83 @@ export class AsyncQueue {
...
+
     return new Promise((resolve) => {
       this.socket.on('connect', () => {
         debug('socket connected');
-        resolve();
+
+        if (sslEnabled) {
+          const requestSSLMessage = Buffer.allocUnsafe(8);
+          requestSSLMessage.writeInt32BE(8, 0);
+          requestSSLMessage.writeInt32BE(80877103, 4);
```
Comment (adelsz) [inline packages/wire/src/queue.ts:55]: Can this message be moved into the messages modules?
https://github.com/adelsz/pgtyped/blob/master/packages/wire/src/messages.ts

**Review by adelsz — COMMENTED** [review]
Thanks for the PR @water-a!
Looks great overall, added a few comments regarding message serialization/parsing

File: packages/wire/src/queue.ts (hunk 2 above)
Comment (water-a) [reply packages/wire/src/queue.ts:55]: Yea I was going to go about doing that, but unsure how to convert it to the patterns array.

I've tried doing patterns: [ int32(8), int(80877103) ], but it adds some extra bits of info in the beginning.

File: packages/wire/src/queue.ts (hunk 1 above)
Comment (water-a) [reply packages/wire/src/queue.ts:77]: So one issue I saw with doing that is I'm not sure how to handle multiple indicator responses with how you have everything setup.

File: packages/wire/src/queue.ts (hunk 2 above)
Comment (adelsz) [reply packages/wire/src/queue.ts:55]: The first 4 bytes is the message length which is added automatically by the message processor, will something like this work?
```
sslRequest: {
    name: 'SSLRequestMessage',
    type: 'CLIENT',
    indicator: null,
    pattern: (data) => [
      // The SSL request code.
      int32(80877103),
    ],
  }
```

File: packages/wire/src/queue.ts (hunk 1 above)
Comment (adelsz) [reply packages/wire/src/queue.ts:77]: Ah, ok. Looking at the docs [here](https://www.postgresql.org/docs/9.3/protocol-flow.html#AEN99727), it does seem like the N/S response isn't a standard pgsql protocol message, so I guess it is ok to leave it as is.

File: packages/wire/src/queue.ts (hunk 2 above)
Comment (water-a) [reply packages/wire/src/queue.ts]: Yup that seems to work!

**Review by adelsz — APPROVED** [review]
LGTM! Thanks @water-a

**Comment by adelsz** [comment]
@water-a One minor request: can you please add a small section about the ssl config param to the [docs](https://pgtyped.now.sh/docs/cli)?

**Comment by water-a** [comment]
> @water-a One minor request: can you please add a small section about the ssl config param to the [docs](https://pgtyped.now.sh/docs/cli)?

Sure thing!

**Pull request merged**